### PR TITLE
New module samtools/sormadup

### PR DESCRIPTION
**Pull request by matthdsm**
Second try at a samtools meta module

xref #3310 
xref #4571 

Speak now or forever hold your peace

## PR checklist

Closes #XXX 

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [ ] Remove all TODO statements.
- [ ] Emit the `versions.yml` file.
- [ ] Follow the naming conventions.
- [ ] Follow the parameters requirements.
- [ ] Follow the input/output options guidelines.
- [ ] Add a resource `label`
- [ ] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - For modules:
    - [ ] `nf-core modules test <MODULE> --profile docker`
    - [ ] `nf-core modules test <MODULE> --profile singularity`
    - [ ] `nf-core modules test <MODULE> --profile conda`
  - For subworkflows:
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile docker`
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile singularity`
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile conda`

### CODE REVIEW

**Comment by mahesh-panchal** [comment]
Needs more `argsN`. See module guidelines. There's some stuff there already.

**Comment by matthdsm** [comment]
Added the `argsN`, but kept the already defined args since those are essential to make the module function

File: modules/nf-core/samtools/sormadup/main.nf
Comment (mahesh-panchal) [inline modules/nf-core/samtools/sormadup/main.nf:72]: > The numbering of each $args variable MUST correspond to the order of the tools, and MUST be documented in meta.yml. E.g. in the first example, bwa mem is the first tool so is given $args, samtools view is the second tool so is $args2, etc.

https://nf-co.re/docs/contributing/modules#new-module-guidelines-and-pr-review-checklist

https://nf-co.re/docs/contributing/modules#documentation

Comment (matthdsm) [reply modules/nf-core/samtools/sormadup/main.nf:72]: Is there an example somewhere of adding `args` to meta.yml? I want to do it right from the get go.

Comment (mahesh-panchal) [reply modules/nf-core/samtools/sormadup/main.nf]: Can search modules for `args_id` in the `meta.yml`s for concrete examples. I'm referring to point 5 in the documentation link I posted above.

Comment (matthdsm) [reply modules/nf-core/samtools/sormadup/main.nf]: Fixed it! Can you check again?

File: modules/nf-core/samtools/sormadup/main.nf
Comment (matthdsm) [reply modules/nf-core/samtools/sormadup/main.nf:35]: I'm aware it isn't ideal, but upping the memory in samtools sort speeds it up tremendously in our tests.
Would you be comfortable with a more conservative 60% instead of 75?

File: modules/nf-core/samtools/sormadup/main.nf
Comment (mahesh-panchal) [reply modules/nf-core/samtools/sormadup/main.nf]: LGTM

File: modules/nf-core/samtools/sormadup/meta.yml
Comment (mahesh-panchal) [inline modules/nf-core/samtools/sormadup/meta.yml:6]: Perhaps add the rest of the tool subcommands to the keyword list

File: modules/nf-core/samtools/sormadup/main.nf
Comment (muffato) [reply modules/nf-core/samtools/sormadup/main.nf:35]: I think I would integrate this module with a patch to remove the requirement. Then do some tests and feed back to you / reinstate the requirement.

Let's just keep this line for now

File: modules/nf-core/samtools/sormadup/meta.yml
Comment (matthdsm) [reply modules/nf-core/samtools/sormadup/meta.yml:9]: Done!

**Review by mahesh-panchal — APPROVED** [review]
Works for me. 

I would say get approval from a maintainer before merging.

**Comment by matthdsm** [comment]
> Works for me.
> 
> I would say get approval from a maintainer before merging.

Thanks! I'd like _several_ approvals before I merge 😅

**Review by muffato — APPROVED** [review]

**Comment by muffato** [comment]
@matthdsm : did you see nf-test is now failing ?

**Comment by matthdsm** [comment]
@muffato, yes, I'm currently checking out why 🤔  Probably something to do with the bam headers

File: modules/nf-core/samtools/sormadup/main.nf
Comment (maxulysse) [inline modules/nf-core/samtools/sormadup/main.nf:33]:
```suggestion
    def extension = args.contains("--output-fmt sam") ? "sam" :
                    args.contains("--output-fmt cram") ? "cram" :
                    "bam"
```

File: modules/nf-core/samtools/sormadup/main.nf
Comment (maxulysse) [inline modules/nf-core/samtools/sormadup/main.nf:88]:
```suggestion
    def extension = args.contains("--output-fmt sam") ? "sam" :
                    args.contains("--output-fmt cram") ? "cram" :
                    "bam"
```

File: modules/nf-core/samtools/sormadup/meta.yml
Comment (maxulysse) [inline modules/nf-core/samtools/sormadup/meta.yml:13]: I'd add another keyword for denoting such a module

**Review by maxulysse — APPROVED** [review]
Looking good, I love this

File: modules/nf-core/samtools/sormadup/meta.yml
Comment (matthdsm) [reply modules/nf-core/samtools/sormadup/meta.yml:13]: suggestions?

Comment (mahesh-panchal) [reply modules/nf-core/samtools/sormadup/meta.yml:13]: Suggestions: 
- `multi-tool` // My pick
-  `optimised`
- `enhanced`
- `performant`

Comment (matthdsm) [reply modules/nf-core/samtools/sormadup/meta.yml:13]: `multi-tool` it is!

Comment (muffato) [reply modules/nf-core/samtools/sormadup/meta.yml:13]: +1 for `multi-tool`

**Pull request merged**
